src/passes/publicFunctionSplitter/externalFunctionCreator.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  Block,
+  ContractDefinition,
+  FunctionDefinition,
+  FunctionVisibility,
+  Expression,
+  Return,
+  FunctionKind,
+} from 'solc-typed-ast';
+import { AST } from '../../ast/ast';
+import { ASTMapper } from '../../ast/mapper';
+import { cloneASTNode } from '../../utils/cloning';
+import { createCallToFunction } from '../../utils/functionStubbing';
+import { createIdentifier } from '../../utils/nodeTemplates';
+export class ExternalFunctionCreator extends ASTMapper {
+  suffix = '_internal';
+
+  constructor(public InternalToExternalFunctionMap: Map<FunctionDefinition, FunctionDefinition>) {
+    super();
+  }
+  /*
+  This class will visit each function definition. If the function definition is public it will
+  create an external counterpart. The visited function will then be changed to internal and have 
+  the suffix _internal added to the name. 
+
+  The internal and external functions will be added to the Maps above to be used in the next pass 
+  to modify the function calls. All internal function calls to the original function will need to 
+  be renamed and all contract to contract calls need to have their referenced changed to the external 
+  function (Still not supported yet). 
+  */
+
+  visitFunctionDefinition(node: FunctionDefinition, ast: AST): void {
+    if (FunctionVisibility.Public === node.visibility && node.kind !== FunctionKind.Constructor) {
+      const newExternalFunction = this.createExternalFunctionDefintion(node, ast);
+      this.insertReturnStatement(node, newExternalFunction, ast);
+      this.modifyPublicFunction(node);
+      this.InternalToExternalFunctionMap.set(node, newExternalFunction);
+    }
+    this.commonVisit(node, ast);
+  }
+
+  private modifyPublicFunction(node: FunctionDefinition): FunctionDefinition {
+    node.visibility = FunctionVisibility.Internal;
+    node.name = node.name + this.suffix;
+    return node;
```

Comment:
As far as I can tell this can be a void function

Reply:
Yeah good spot.